packages/wallet/notes/action-routing.md

```diff
@@ -0,0 +1,91 @@
+# Action Routing
+
+## The Action Routing Problem
+
+If we have multiple protocols running in the wallet simultaneously, we need a way of identifying
+which protocol a given action should apply to. This applies to the majority of our actions,
+including messages from an opponent, transaction updates from the blockchain or UI interaction
+by the user.
+
+We model each protocol as running inside a given process, which is identifiable by a `processId`.
+These protocols can be constructed from other protocols, so it is possible to have a 'parent protocol'
+running inside a process, with several 'child protocols' running (perhaps simultaneously)
+inside it. We will refer to the set of nested protocols running at any one time inside a
+single process as the _protocol tree_.
+
+## Current proposal: `processId`
+
+The current proposal is to include a `processId` on all actions that should be routed to an
+existing protocol. We call these 'existing-process' actions. A top-level router would
+forward these on to the relevant process, which would then forward them on to child
+protocols if appropriate.
+
+## Problems with the current proposal
+
+The current proposal gives a way for actions to be routed to the correct process but _doesn't_
+provide any information as to how to route them further.
+
+As a concrete example, suppose we have a protocol for withdrawing from a channel nested inside
+a ledger channel. It's possible that we would want to attempt to submit a couple of blockchain
+transactions in parallel (perhaps from separate accounts to get around nonce issues). When
+the process receives a `TransactionConfirmed` event, it sends it to the parent protocol. The
+parent protocol needs to somehow decide which of the sub-protocols this event applies to.
+
+As another example, suppose we have a protocol for responding to a challenge which wants to
+handle the case where you respond to the challenge but then receive a state that you could
+refute with from your opponent. The protocol wants to allow the user to submit a refute transaction
+with the same nonce and a higher gas amount, hoping to replace the original response. Again,
+it is important for the protocol to be able to understand which of the two transaction-submitted
+sub-protocols a `TransactionConfirmed` event applies to in this case.
+
+## Proposed Solution
+
+Here's one possible solution to this problem:
+
+1. We replace the `processId` with a `protocolLocator` string.
+2. The `protocolLocator` would look something like `process-123.refute-transaction.attempt-1`.
```

Comment:
So, I think I would suggest having
1. a `processId` on every action that's meant to be routed to a process
2. maybe a `processLocator` (naming?) that is an _array_ of strings, rather than a `.`-delimited string